Make MockGamepadProvider robust so that it doesn't end up crashing the test runner

chirags27 opened the pull request:
#### feab6de8fb5ee04b52a859ee7066266258c13b30
<pre>
Make MockGamepadProvider robust so that it doesn&apos;t end up crashing the test runner
<a href="https://bugs.webkit.org/show_bug.cgi?id=250605">https://bugs.webkit.org/show_bug.cgi?id=250605</a>
rdar://104165858

Reviewed by David Kilzer.

This change adds more state to MockGamepadProvider so that it can track the gamepads
that were connected before the GamepadProviderClient starts monitoring. Before this change,
a simple snippet like:
&lt;script&gt;
  testRunner.setMockGamepadDetails(0);
  testRunner.connectMockGamepad(0);
  addEventListener(&apos;gamepadconnected&apos;, () =&gt; {});
&lt;/script&gt;

would crash the test runner.

* Source/WebCore/testing/MockGamepadProvider.cpp:
(WebCore::MockGamepadProvider::startMonitoringGamepads):
(WebCore::MockGamepadProvider::stopMonitoringGamepads):
(WebCore::MockGamepadProvider::connectMockGamepad):
(WebCore::MockGamepadProvider::disconnectMockGamepad):
* Source/WebCore/testing/MockGamepadProvider.h:

Canonical link: <a href="https://commits.webkit.org/259411@main">https://commits.webkit.org/259411@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44a3ee8ef441ec3d7fa28ebc10c0c468177a05e3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104797 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13877 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37698 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114070 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174272 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108707 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15007 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4804 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97127 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112984 "Built successfully") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11599 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94611 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39125 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93459 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26225 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7226 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27585 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7325 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4177 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13377 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47135 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6488 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9112 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/testing/MockGamepadProvider.h:
```diff
@@ -59,6 +61,7 @@ class MockGamepadProvider : public GamepadProvider {
     void gamepadInputActivity();
 
     Vector<PlatformGamepad*> m_connectedGamepadVector;
+    HashMap<GamepadProviderClient*, HashSet<PlatformGamepad*>>  m_invisibleGamepadsForClient;
```
Please use WeakHashMap and WeakHashSet instead

chirags27 replied:
@rniwa The classes in question here aren't subclasses of `CanMakeWeakPtr `, so I don't think these can be used in the WeakMap or Set directly, no? If we were to do that, it should probably be a separate radar?

chirags27 replied:
Also, a lot of other classes track these objects in non weak containers, for example, see: GameControllerGamepadProvider

So we should probably defer moving to weak until we can migrate everything

rniwa replied:
Doing it in a separate PR sounds good to me.

ddkilzer commented on Source/WebCore/testing/MockGamepadProvider.h:
```diff
@@ -59,6 +61,7 @@ class MockGamepadProvider : public GamepadProvider {
     void gamepadInputActivity();
 
     Vector<PlatformGamepad*> m_connectedGamepadVector;
+    HashMap<GamepadProviderClient*, HashSet<PlatformGamepad*>>  m_invisibleGamepadsForClient;
```
Per the earlier conversion with Ryosuke about using a `WeakPtr<>` here, please add a `FIXME` comment in case you get busy with other tasks so that someone else can make the change later.  I would explicitly spell out which raw pointers should be changed to weak pointers by restating the type, perhaps like this if this is correct:

```
    HashMap<GamepadProviderClient*, HashSet<PlatformGamepad*>>  m_invisibleGamepadsForClient;  // FIXME: Use HashMap<WeakPtr<GamepadProviderClient>, HashSet<WeakPtr<PlatformGamepad>>>.
```

ddkilzer commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -106,8 +119,14 @@ bool MockGamepadProvider::disconnectMockGamepad(unsigned index)
+            auto setOfGamepads = m_invisibleGamepadsForClient.get(client);
+            setOfGamepads.remove(m_mockGamepadVector[index].get());
+            m_invisibleGamepadsForClient.set(client, setOfGamepads);
+        } else
```
Would be nice to pull out `m_mockGamepadVector[index].get()` into a well-named variable outside the `for` loop since it apparently never changes inside the loop.

Also, I think this variable should be `auto& setOfGamepads` so we don't make an unwanted copy of the `HashSet`, correct?  Then we don't have to call `m_invisibleGamepadsForClient.set(client, setOfGamepads);` at all.  Or is there a specific reason to make a full copy of the `HashSet`, modify it, then replace the original?

chirags27 replied:
Makes sense, the copy wasn't intentional there!

chirags27 replied:
Fixed using `find`

ddkilzer commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -87,8 +94,14 @@ bool MockGamepadProvider::connectMockGamepad(unsigned index)
+    for (auto& client : m_clients) {
         client->platformGamepadConnected(*m_connectedGamepadVector[index], EventMakesGamepadsVisible::Yes);
+        if (m_invisibleGamepadsForClient.contains(client)) {
+            for (auto invisibleGamepad : m_invisibleGamepadsForClient.get(client))
```
Should this be `auto& invisibleGamepad` inside in the `for` loop to prevent unneeded copies of the `HashSet`, or does `client->platformGamepadConnected()` modify the `HashSet` that it is passed?  (If it doesn't modify the `HashSet`, we should be able to use `auto&` here.)

ddkilzer commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -48,12 +48,19 @@ void MockGamepadProvider::startMonitoringGamepads(GamepadProviderClient& client)
+        if (gamepad)
+            invisibleGamepads.add(gamepad);
+    }
+    m_invisibleGamepadsForClient.set(&client, invisibleGamepads);
```
Maybe add an assertion here to match what is done for `m_clients` above:

```
    ASSERT(!m_invisibleGamepadsForClient.contains(&client));
```

ddkilzer reviewed (changes requested):
Please make the requested changes.  Otherwise, this patch looks good!

ddkilzer commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -106,8 +120,14 @@ bool MockGamepadProvider::disconnectMockGamepad(unsigned index)
+    for (auto& client : m_clients) {
+        if (m_invisibleGamepadsForClient.contains(client) && m_invisibleGamepadsForClient.get(client).contains(gamepadToRemove)) {
+            auto itr = m_invisibleGamepadsForClient.find(client);
+            itr->value.remove(gamepadToRemove);
```
I know this is really nit-picky, but please use `auto iter` instead of `auto itr` here as the former is much more commonly used in WebKit project sources.

You could also combine the two lines into a single line:
```
             m_invisibleGamepadsForClient.find(client)->value.remove(gamepadToRemove);
```

ddkilzer reviewed (approved):
r=me, but please fix the `auto itr` variable name (or remove it).

ddkilzer commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -106,8 +120,13 @@ bool MockGamepadProvider::disconnectMockGamepad(unsigned index)
+    for (auto& client : m_clients) {
+        if (m_invisibleGamepadsForClient.contains(client) && m_invisibleGamepadsForClient.get(client).contains(gamepadToRemove)) {
+            m_invisibleGamepadsForClient.find(client)->value.remove(gamepadToRemove);
+        } else
```
[Style](https://ews-build.webkit.org/#/builders/6/builds/104751) bot says we can remove braces here.

The `if` expression is long enough that I'm okay leaving the braces in place, though we should try to catch this in the future.

ddkilzer replied:
Merge queue seems to have failed, so we should fix the style issue.

cdumez commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -87,8 +95,14 @@ bool MockGamepadProvider::connectMockGamepad(unsigned index)
+    for (auto& client : m_clients) {
         client->platformGamepadConnected(*m_connectedGamepadVector[index], EventMakesGamepadsVisible::Yes);
+        if (m_invisibleGamepadsForClient.contains(client)) {
+            for (auto& invisibleGamepad : m_invisibleGamepadsForClient.get(client))
```
This is testing code so I don't care too much but this is inefficient. We're doing a hash map lookup 3 times here. Once in contains(), then in get() and once again in remove().

cdumez commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -106,8 +120,13 @@ bool MockGamepadProvider::disconnectMockGamepad(unsigned index)
-        client->platformGamepadDisconnected(*m_mockGamepadVector[index]);
+    auto gamepadToRemove = m_mockGamepadVector[index].get();
+    for (auto& client : m_clients) {
+        if (m_invisibleGamepadsForClient.contains(client) && m_invisibleGamepadsForClient.get(client).contains(gamepadToRemove))
```
Ditto.

cdumez commented on Source/WebCore/testing/MockGamepadProvider.h:
```diff
@@ -59,6 +61,9 @@ class MockGamepadProvider : public GamepadProvider {
     void gamepadInputActivity();
 
     Vector<PlatformGamepad*> m_connectedGamepadVector;
+    // FIXME: Use HashMap<WeakPtr<GamepadProviderClient>, HashSet<WeakPtr<PlatformGamepad>>>
```
Adding new code with raw pointers in containers like this is really scary, even with a FIXME. This goes against the direction we want to take our code towards.

chirags27 replied:
I agree, but there are other places in the production code which do the same thing today because these classes don't support WeakPtr.

I think we should have a separate change so that `GamepadProviderClient` and `PlatformGamepad` support WeakPtr, and change all these places?

cdumez replied:
Ok.

ddkilzer replied:
Previously discussed here with @rniwa:  <https://github.com/WebKit/WebKit/pull/8648#discussion_r1072974143>

cdumez commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -87,8 +95,15 @@ bool MockGamepadProvider::connectMockGamepad(unsigned index)
+        if (gamepadsForClient != m_invisibleGamepadsForClient.end()) {
+            for (auto& invisibleGamepad : gamepadsForClient->value)
+                client->platformGamepadConnected(*invisibleGamepad, EventMakesGamepadsVisible::Yes);
+            m_invisibleGamepadsForClient.remove(client);
```
Seems like this should be `remove(gamepadsForClient)` to avoid an extra hash map look up.

cdumez commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -106,8 +121,14 @@ bool MockGamepadProvider::disconnectMockGamepad(unsigned index)
+    auto gamepadToRemove = m_mockGamepadVector[index].get();
+    for (auto& client : m_clients) {
+        auto gamepadsForClient = m_invisibleGamepadsForClient.find(client);
+        if (gamepadsForClient != m_invisibleGamepadsForClient.end() && gamepadsForClient->value.contains(gamepadToRemove))
```
Similarly here, the double look up of gamepadToRemove in contains() and remove() is still unfortunate.

This could we written as:
```
if (gamepadsForClient == m_invisibleGamepadsForClient.end() || !gamepadsForClient->value.remove(gamepadToRemove))
    client->platformGamepadDisconnected(*m_mockGamepadVector[index]);

chirags27 replied:
Good point

chirags27 committed:
Make MockGamepadProvider robust so that it doesn't end up crashing the test runner
https://bugs.webkit.org/show_bug.cgi?id=250605
rdar://104165858

Reviewed by David Kilzer.

This change adds more state to MockGamepadProvider so that it can track the gamepads
that were connected before the GamepadProviderClient starts monitoring. Before this change,
a simple snippet like:
<script>
  testRunner.setMockGamepadDetails(0);
  testRunner.connectMockGamepad(0);
  addEventListener('gamepadconnected', () => {});
</script>

would crash the test runner.

* Source/WebCore/testing/MockGamepadProvider.cpp:
(WebCore::MockGamepadProvider::startMonitoringGamepads):
(WebCore::MockGamepadProvider::stopMonitoringGamepads):
(WebCore::MockGamepadProvider::connectMockGamepad):
(WebCore::MockGamepadProvider::disconnectMockGamepad):
* Source/WebCore/testing/MockGamepadProvider.h:

Canonical link: https://commits.webkit.org/259411@main
